### DISCUSSION TITLE
using different seeds in each part of the trajectory

### DISCUSSION THREAD

**Discussion by Arash-72** [r-simmer/simmer#306] [Q&A] [answered]
Hi,
I want to simulate a clinic in which a patient upon arrival enters a room (if at least there is one empty room available) and then visits a nurse for a random time duration and then visits a doctor for a random time duration.
Assume that I know the arrival time of all patients in a day and I want to see how much changing the number of exam rooms changes the patients' wait time. But to have a fair comparison, the nurse visit durations (also the doctor visit durations) should be the same in all scenarios.  I mean if there are 4 patients in the schedule of that day and if the nurse visit durations for those 4 patients are 2.5, 3.5, 3.6, and 2.8 for the scenario with 2 rooms, I need to get the same random durations (2.5, 3.5, 3.6, and 2.8) for the scenario with 3 rooms (also the same story for doctor visit duration). 

Right now my code is like 

```r
set.seed(100)
TRJ <- trajectory() %>%
seize("room", 1) %>% 
seize("nurse", 1) %>%
timeout(function() rgamma(1, 4, 2))  %>%
release("nurse", 1) %>%
seize("doctor", 1) %>%
timeout(function() rgamma(1, 5, 3))  %>%
release("doctor", 1) %>%
release("room", 1) 
```

However, once I change the number of rooms from 3 to 4, the nurse visit durations (and doctor visit durations) don't remain the same, while I need them to remain the same. 
Is it possible to use a certain seed inside the timeout() function? Or define different streams of seeds and let them be used for each type of resource? For example, having a certain seed for the nurse visit duration while having another seed for the doctor visit duration?

Thank you so much for your time and help in advance.

**Comment by Enchufa2** [comment, marked as answer]
You can set a dataframe with predefined times, which are added as attributes, and then read them from there:

```r
library(simmer)
set.seed(100)

patient <- trajectory() %>%
  seize("room", 1) %>% 
  seize("nurse", 1) %>%
  timeout_from_attribute("nurse") %>%
  release("nurse", 1) %>%
  seize("doctor", 1) %>%
  timeout_from_attribute("doctor") %>%
  release("doctor", 1) %>%
  release("room", 1)

n <- 5 # number of patients
df <- data.frame(
  time = rexp(n, 1/10), # your interarrival times
  nurse = rgamma(n, 2, 2),
  doctor = rgamma(n, 5, 3)
)

simmer() %>%
  add_resource("room", 10) %>%
  add_resource("nurse", 2) %>%
  add_resource("doctor", 2) %>%
  add_dataframe("patient", patient, df) %>%
  run()
```

**Comment by Arash-72** [comment]
Thank you so much for your quick answer. That is a great help.

Unfortunately, there is still an issue. In my simulation trajectory, there is also a branch. For example, assume that patients are required to visit the nurse by 40% probability. So it is like:
set.seed(100)
TRJ <- trajectory() %>%
  branch( function() runif(1) < 0.40, continue = c(TRUE, TRUE),
          trajectory() %>% 
            seize("room", 1) %>% 
            seize("nurse", 1) %>%
            timeout(function() rgamma(1, 4, 2))  %>%
            release("nurse", 1),
          trajectory() %>% 
            seize("room", 1)
  ) %>%
  seize("doctor", 1) %>%
  timeout(function() rgamma(1, 5, 3))  %>%
  release("doctor", 1) %>%
  release("room", 1) 

I was wondering if you could also let me know how I can make the same patients visit the nurse in each scenario. For example if in the scenario with 2 rooms patients 1 and 4 visit the nurse how I can make only patients 1 and 4 visit the nurse in the scenario with 3 rooms?  I followed an approach similar to yours as below:

Position <- function(n){
  vec <- vector()
  for (i in 1:n) {
    rand = runif(1)
    if (rand < 0.4) {
      vec[i] <- 1  # First sub-trajectory
    } else {
      vec[i] <- 2
    } 
  }
  return(vec)
}


library(simmer)
set.seed(100)


env <- simmer()
patient <- trajectory() %>%
  seize("room", 1) %>% 
  branch("Branch",
         continue = c(TRUE, TRUE),
         trajectory() %>%
           seize("nurse", 1) %>%
           timeout_from_attribute("nurse") %>%
           release("nurse", 1),
         trajectory() 
         ) %>%
  seize("doctor", 1) %>%
  timeout_from_attribute("doctor") %>%
  release("doctor", 1) %>%
  release("room", 1)

n <- 5 # number of patients
df <- data.frame(
  time = rexp(n, 1/10), # your interarrival times
  nurse = rgamma(n, 2, 2),
  doctor = rgamma(n, 5, 3),
  Branch = Position(n)
)

env %>%
  add_resource("room", 10) %>%
  add_resource("nurse", 1) %>%
  add_resource("doctor", 1) %>%
  add_dataframe("patient", patient, df, mon = 2) %>%
  run()

However, it doesn't work.

I would greatly appreciate your help on this in advance.

**Reply by Enchufa2** [reply]
Same approach:

1. add another column with TRUE/FALSE;
2. this will be added as an attribute;
3. read that attribute in your branching function.

**Reply by Arash-72** [reply]
Thank you very much. I defined the attribute as below:
Position <- function(n){
  vec <- vector()
  for (i in 1:n) {
    rand = runif(1)
    if (rand < 0.4) {
      vec[i] <- TRUE  # First sub-trajectory
    } else {
      vec[i] <- FALSE
    } 
  }
  return(vec)
}

n<- 5
df <- data.frame(
  time = rexp(n, 1/10), # your interarrival times
  nurse = rgamma(n, 2, 2),
  doctor = rgamma(n, 5, 3),
  Branch = Position(n)
)

And for the branch function, I tried both below ways but none of them didn't work.
1:
branch(function(){df$Branch},
         continue = c(TRUE, TRUE), ..
) %>% ...

2:
branch(function() {"Branch"},
         continue = c(TRUE, TRUE), ...
) %>%...

Could you please show me what the function that reads the attribute from df should look like?

**Reply by Enchufa2** [reply]
The get_attribute function is what you are looking for.

**Reply by Arash-72** [reply]
Perfect!
Thank you so much!🙏🙏
